src/persist/src/mem.rs

```diff
@@ -167,8 +168,10 @@ impl Meta for MemStream {
 }
 
 /// An in-memory implementation of [Snapshot].
+#[derive(Debug, Eq, PartialEq)]
```

Comment:
this Eq,PartialEq and making dataz public weakens the abstraction in an easily avoided way, so my preference is for the test to extract data via the Snapshot interface. but leave it if you feel strongly otherwise. apologies if this was here the last time around and i missed it

```
        let snapshot = file_stream.snapshot().expect("fixme");
        let actual = Vec::new();
        while snapshot.read(&mut actual) {};
        assert_eq!(actual, data);
```

if we wanted to make this a one liner (which could be worthwhile, we're probably gonna do this a lot. i have something similar in my PR), we could make a trait implemented for anything that implements Snapshot

```
trait Extract<T> {
  fn extract(&mut self) -> Vec<T>;
}
impl<S: Snapshot> Extract<(String, String), u64, isize)> for S {
  fn extract(&mut self) -> Vec<T> {
    let mut buf = Vec::new();
    while self.read(&mut buf) {};
    buf
  }
}
```

(Debug is basically always fine IMO)

Reply:
do you have a preference between adding a separate extract trait, vs providing a `read_to_end` function with a default impl to the snapshot trait

the function could look like this (within the definition of snapshot): 

```
    /// A full read of the data in the snapshot.
    fn read_to_end<E: Extend<((String, String), u64, isize)>>(&mut self, buf: &mut E) {
        while self.read(buf) {}
    }
 ```
 and the test code would be like
 ```
         let mut snapshot = file_stream.snapshot()?;
        let mut snapshot_data = Vec::new();
        snapshot.read_to_end(&mut snapshot_data);
        assert_eq!(snapshot_data, data);
        ```
        
let me know what you think of that

Reply:
my thought process here is that I think being able to read everything will probably be useful outside of tests even and theres some precedent for this kind of an api in `std::io::Read` but am curious how you think about it

Reply:
Yeah that's nice. I don't think non-test code will want to use it though because snapshots could be large, which is the reason that `read` itself doesn't just have these `read_to_end` semantics. I also suspect there will be quite a few extra `Vec::new`s cluttering up the tests. Let's skip the Extract trait and put something in Snapshot that has the extract signature (returning a Vec instead of taking a mut ref). For good measure, throw a #[cfg(test)] on it, which we can always remove later if I'm wrong about non-test code wanting this.

Reply:
done!